fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -152,21 +150,28 @@ where
     async fn deliver(
         &self,
         mut state: Self::State,
-        mut msg: Self::Message,
+        msg: Self::Message,
     ) -> anyhow::Result<(Self::State, Self::DeliverOutput)> {
         let (apply_ret, emitters, latency) = if msg.from == system::SYSTEM_ACTOR_ADDR {
             let (execution_result, latency) = measure_time(|| state.execute_implicit(msg.clone()));
             let (apply_ret, emitters) = execution_result?;
 
             (apply_ret, emitters, latency)
         } else {
-            // TODO: maybe compare the gas limits is better?
-            msg.gas_limit = msg.gas_limit.min(self.gas.available_block_gas());
+            if msg.gas_limit > state.gas_market().available_block_gas() {
+                tracing::warn!("gas limit exceed available block gas limit");
+            }
 
             let (execution_result, latency) = measure_time(|| state.execute_explicit(msg.clone()));
             let (apply_ret, emitters) = execution_result?;
 
-            self.gas.consume_gas(apply_ret.msg_receipt.gas_used)?;
+            if state
+                .gas_market_mut()
+                .record_gas_used(apply_ret.msg_receipt.gas_used)
+                .is_err()
+            {
+                tracing::warn!("should not have exceeded block gas limit");
```

Comment:
This is poor error handling logic. This code is making an assumption about what the nature of the error was, which may not hold up in the future. Remember: the GasMarket is customizable. Instead, this warning should print the inner error instead:

```suggestion
                tracing::warn!("[ASSERTION FAILED] gas market failed while recording utilization: {err}");
```